lib/prelude/src/lib.rs

```diff
@@ -74,7 +87,119 @@ pub type PhantomData9<T1,T2,T3,T4,T5,T6,T7,T8,T9> = PhantomData2<PhantomData8<T1
 /// ```
 pub fn with<T, F: FnOnce(T) -> Out, Out>(t: T, f: F) -> Out { f(t) }
 
+pub trait OptionOps {
+    type Item;
+    fn for_each<U, F: FnOnce(Self::Item) -> U>(self, f: F);
+}
+
+impl<T> OptionOps for Option<T> {
+    type Item = T;
+    fn for_each<U, F: FnOnce(Self::Item) -> U>(self, f: F) {
+        match self {
+            Some(x) => { f(x); }
+            None    => {}
+        }
+    }
+}
+
+
+
+
+// ===================
+// === WithPhantom ===
+// ===================
+
+/// A wrapper adding a phantom type to a structure.
+#[derive(Derivative)]
+#[derive(Shrinkwrap)]
+#[shrinkwrap(mutable)]
+#[derivative(Clone   (bound="T:Clone"))]
+#[derivative(Default (bound="T:Default"))]
+#[derivative(Debug   (bound="T:Debug"))]
+pub struct WithPhantom<T, P=()> {
+    #[shrinkwrap(main_field)]
+    pub without_phantom: T,
+    phantom: PhantomData<P>
+}
+
+impl<T, P> WithPhantom<T, P> {
+    pub fn new(without_phantom: T) -> Self {
+        let phantom = PhantomData;
+        Self { without_phantom, phantom }
+    }
+}
+
+
+// =====================
+// === Rc Extensions ===
+// =====================
+
+/// Using `clone` for structures which are newtype-wrappers over `Rc` is error
+/// prone and hides the real intention. Cloning should always be considered
+/// pricey. This trait adds `clone_rc` method to every such wrapper.
+pub trait IsRc {
+    fn clone_rc(&self) -> Self;
+}
+
+impl<T,S> IsRc for T
+    where T: From<Rc<S>>, T:Deref<Target=Rc<S>> {
+    fn clone_rc(&self) -> Self {
+        Rc::clone(self).into()
+    }
+}
+
+/// See the documentation of `IsRc`. Unfortunately, it's not easy to merge
+/// these two traits together in current Rust version, so `clone_rc` method
+/// for `Rc<S>` is provided separately.
+pub trait RcOps {
+    fn clone_rc(&self) -> Self;
+}
+
+impl<T> RcOps for Rc<T> {
+    fn clone_rc(&self) -> Self {
+        Rc::clone(self)
+    }
+}
+
+
+
+// ==========================
+// === RefCell Management ===
+// ==========================
 
+// === RefGuard ===
+
+pub struct RefGuard<'t,Base,Data> {
```

Comment:
So we can use this in that "SharedDirtyIterator", right?

Reply:
Yes, it seems like this is a safe generalization of that pattern. I haven't still used it much so I cannot guarantee it works in all cases.

Reply:
So let write unit tests for it!